File: math_parser.py
import sympy as sp
from sympy import symbols, sympify, SympifyError
import numpy as np
import re

class MathParser:
    """
    Advanced mathematical expression parser supporting various equation formats
    """
    
    def __init__(self):
        self.x, self.y, self.z, self.t = symbols('x y z t')
        self.common_functions = {
            'sin': sp.sin,
            'cos': sp.cos,
            'tan': sp.tan,
            'exp': sp.exp,
            'log': sp.log,
            'ln': sp.log,
            'sqrt': sp.sqrt,
            'abs': sp.Abs,
            'pi': sp.pi,
            'e': sp.E
        }
    
    def parse_equation(self, equation_str):
        """
        Parse a mathematical equation string into a SymPy expression
        
        Args:
            equation_str (str): Mathematical equation as string
            
        Returns:
            sympy.Expr: Parsed expression or None if invalid
        """
        try:
            # Clean and prepare the equation string
            cleaned_eq = self._clean_equation(equation_str)
            
            # Handle different equation formats
            if '=' in cleaned_eq:
                # Handle equations with explicit z = f(x,y) or y = f(x) format
                left, right = cleaned_eq.split('=', 1)
                if left.strip() in ['z', 'y', 'f']:
                    expression = right.strip()
                else:
                    # Assume it's an implicit equation, move everything to one side
                    expression = f"({left}) - ({right})"
            else:
                expression = cleaned_eq
            
            # Replace common mathematical notation
            expression = self._replace_math_notation(expression)
            
            # Parse with SymPy
            parsed_expr = sympify(expression, locals=self.common_functions)
            
            return parsed_expr
            
        except (SympifyError, ValueError, TypeError) as e:
            print(f"Parsing error: {e}")
            return None
    
    def _clean_equation(self, equation_str):
        """Clean and standardize equation string"""
        # Remove whitespace
        cleaned = equation_str.strip()
        
        # Replace common math symbols
        replacements = {
            '^': '**',  # Power notation
            '×': '*',   # Multiplication
            '÷': '/',   # Division
            '−': '-',   # Minus sign
        }
        
        for old, new in replacements.items():
            cleaned = cleaned.replace(old, new)
        
        return cleaned
    
    def _replace_math_notation(self, expression):
        """Replace mathematical notation with SymPy-compatible syntax"""
        # Handle implicit multiplication (e.g., 2x -> 2*x)
        expression = re.sub(r'(\d)([a-zA-Z])', r'\1*\2', expression)
        expression = re.sub(r'([a-zA-Z])(\d)', r'\1*\2', expression)
        expression = re.sub(r'(\))([a-zA-Z\(])', r'\1*\2', expression)
        expression = re.sub(r'([a-zA-Z\)])(\()', r'\1*\2', expression)
        
        # Handle function notation
        function_replacements = {
            'arcsin': 'asin',
            'arccos': 'acos',
            'arctan': 'atan',
            'sinh': 'sinh',
            'cosh': 'cosh',
            'tanh': 'tanh',
        }
        
        for old, new in function_replacements.items():
            expression = expression.replace(old, new)
        
        return expression
    
    def validate_expression(self, expression):
        """
        Validate if an expression is mathematically sound
        
        Args:
            expression (sympy.Expr): Expression to validate
            
        Returns:
            dict: Validation results with status and messages
        """
        validation_result = {
            'valid': True,
            'warnings': [],
            'errors': []
        }
        
        try:
            # Check for undefined symbols
            free_symbols = expression.free_symbols
            expected_symbols = {self.x, self.y, self.z, self.t}
            unexpected_symbols = free_symbols - expected_symbols
            
            if unexpected_symbols:
                validation_result['warnings'].append(
                    f"Unexpected symbols found: {unexpected_symbols}"
                )
            
            # Check for potential division by zero
            if expression.has(1/self.x) or expression.has(1/self.y):
                validation_result['warnings'].append(
                    "Expression contains division by variables - check for singularities"
                )
            
            # Test evaluation at a few points
            test_points = [(0, 0), (1, 1), (-1, -1), (0.5, 0.5)]
            for point in test_points:
                try:
                    if len(free_symbols) >= 2:
                        result = expression.subs([(self.x, point[0]), (self.y, point[1])])
                    elif len(free_symbols) == 1:
                        result = expression.subs(list(free_symbols)[0], point[0])
                    
                    # Check if result is numeric
                    complex_result = complex(result.evalf())
                    if not np.isfinite(complex_result.real) or not np.isfinite(complex_result.imag):
                        validation_result['warnings'].append(
                            f"Expression may be undefined at point {point}"
                        )
                except Exception as e:
                    validation_result['warnings'].append(
                        f"Evaluation error at point {point}: {str(e)}"
                    )
            
        except Exception as e:
            validation_result['valid'] = False
            validation_result['errors'].append(f"Validation error: {str(e)}")
        
        return validation_result
    
    def extract_domain_info(self, expression):
        """
        Extract domain information from expression
        
        Args:
            expression (sympy.Expr): Expression to analyze
            
        Returns:
            dict: Domain information
        """
        domain_info = {
            'variables': list(expression.free_symbols),
            'restrictions': [],
            'suggested_range': {'x': (-10, 10), 'y': (-10, 10)}
        }
        
        # Check for logarithmic functions
        if expression.has(sp.log):
            domain_info['restrictions'].append("Logarithmic functions require positive arguments")
        
        # Check for square roots
        if expression.has(sp.sqrt):
            domain_info['restrictions'].append("Square root functions require non-negative arguments")
        
        # Check for rational functions
        denominators = []
        for term in sp.preorder_traversal(expression):
            if term.is_Pow and term.exp.is_negative:
                denominators.append(term.base)
        
        if denominators:
            domain_info['restrictions'].append(
                f"Avoid zeros in denominators: {denominators}"
            )
        
        return domain_info
    
    def convert_to_lambda(self, expression, variables=None):
        """
        Convert SymPy expression to lambda function for numerical evaluation
        
        Args:
            expression (sympy.Expr): Expression to convert
            variables (list): List of variables in order
            
        Returns:
            function: Lambda function for numerical evaluation
        """
        if variables is None:
            free_symbols = list(expression.free_symbols)
            # Sort symbols by name for consistency
            variables = sorted(free_symbols, key=str)
        
        try:
            return sp.lambdify(variables, expression, 'numpy')
        except Exception as e:
            print(f"Lambda conversion error: {e}")
            return None

File: src/utils/mathParser.ts
import { evaluate, parse, MathNode } from 'mathjs';

export interface ParsedExpression {
  node: MathNode;
  variables: string[];
  isValid: boolean;
  errorMessage?: string;
}

export function parseMathExpression(equation: string): ParsedExpression {
  try {
    // Clean the equation string
    let cleanEquation = equation.trim();
    
    // Handle equation format (z = f(x,y) or f(x,y))
    if (cleanEquation.includes('=')) {
      const parts = cleanEquation.split('=');
      if (parts.length === 2) {
        const left = parts[0].trim();
        const right = parts[1].trim();
        if (left === 'z' || left === 'y' || left === 'f') {
          cleanEquation = right;
        } else {
          // Assume implicit equation, move to one side
          cleanEquation = `(${left}) - (${right})`;
        }
      }
    }
    
    // Replace common mathematical notation
    cleanEquation = cleanEquation
      .replace(/\^/g, '^')  // Power
      .replace(/×/g, '*')   // Multiplication
      .replace(/÷/g, '/')   // Division
      .replace(/−/g, '-')   // Minus
      .replace(/π/g, 'pi')  // Pi
      .replace(/sin⁻¹/g, 'asin')
      .replace(/cos⁻¹/g, 'acos')
      .replace(/tan⁻¹/g, 'atan')
      .replace(/ln/g, 'log')
      .replace(/√/g, 'sqrt');
    
    // Handle implicit multiplication
    cleanEquation = cleanEquation
      .replace(/(\d)([a-zA-Z])/g, '$1*$2')  // 2x -> 2*x
      .replace(/([a-zA-Z])(\d)/g, '$1*$2')  // x2 -> x*2
      .replace(/\)([a-zA-Z(])/g, ')*$1')    // )(x -> )*(x
      .replace(/([a-zA-Z])\(/g, '$1*(');    // x( -> x*(
    
    // Parse the expression
    const node = parse(cleanEquation);
    
    // Extract variables
    const variables = extractVariables(node);
    
    return {
      node,
      variables,
      isValid: true
    };
    
  } catch (error) {
    return {
      node: parse('0'), // fallback
      variables: [],
      isValid: false,
      errorMessage: error instanceof Error ? error.message : 'Unknown parsing error'
    };
  }
}

function extractVariables(node: MathNode): string[] {
  const variables = new Set<string>();
  
  node.traverse((node, path, parent) => {
    if (node.type === 'SymbolNode') {
      const name = (node as any).name;
      // Only include single-letter variables and common math variables
      if (/^[a-zA-Z]$/.test(name) || ['pi', 'e'].includes(name)) {
        if (!['pi', 'e', 'sin', 'cos', 'tan', 'log', 'sqrt', 'exp', 'abs', 'asin', 'acos', 'atan', 'sinh', 'cosh', 'tanh'].includes(name)) {
          variables.add(name);
        }
      }
    }
  });
  
  return Array.from(variables).sort();
}

export function evaluateExpression(
  parsedExpr: ParsedExpression, 
  variables: Record<string, number>
): number {
  try {
    if (!parsedExpr.isValid) {
      return NaN;
    }
    
    const result = parsedExpr.node.evaluate(variables);
    return typeof result === 'number' ? result : NaN;
  } catch (error) {
    return NaN;
  }
}

export function createMeshData(
  parsedExpr: ParsedExpression,
  xMin: number,
  xMax: number,
  yMin: number,
  yMax: number,
  resolution: number
): { x: number[][], y: number[][], z: number[][] } {
  const x: number[][] = [];
  const y: number[][] = [];
  const z: number[][] = [];
  
  const xStep = (xMax - xMin) / (resolution - 1);
  const yStep = (yMax - yMin) / (resolution - 1);
  
  for (let i = 0; i < resolution; i++) {
    x[i] = [];
    y[i] = [];
    z[i] = [];
    
    for (let j = 0; j < resolution; j++) {
      const xVal = xMin + j * xStep;
      const yVal = yMin + i * yStep;
      
      x[i][j] = xVal;
      y[i][j] = yVal;
      
      const zVal = evaluateExpression(parsedExpr, { x: xVal, y: yVal });
      z[i][j] = isFinite(zVal) ? zVal : null;
    }
  }
  
  return { x, y, z };
}